### PR TITLE
Generate specs from the output of solc-to-k

### PR DESCRIPTION
**Pull request by ehildenb**
This PR enables scraping a K definition for contract names and producing specification templates for it.

- function `binRuntime(_)` is declared in central place to support multiple contracts. It's declared over the `Contract` sort, which `solc-to-k` now extends with subsorts.
- `kevm solc-to-k` now outputs `solc` stderr on failure to compile.
- `kevm solc-to-k` adds `symbol` attribute to more productions, and generates a better token grammar.
- Functions `abstract_cell_vars`, `get_productions`, `get_production_for_klabel`, `get_label_for_cell_sorts`, and `build_empty_configuration_cell` are pulled in as utilities from ERC20 tool.
- `kevm gen-spec` command is added, which takes a kompiled directory and generates generic "template" specs, one per contract.
- Because the default module for Haskell backend is shifted to `EDSL`, some tests that were for executing tests/programs do not work with haskell backend anymore. They were not useful tests at the moment anyway, so I removed them.

The added test in `tests/gen-spec` tests that given a bunch of files generated with `kevm solc-to-k` (these file are committed to the repo), can we call `kevm gen-spec ...` and it produces the correct output? (it correctly scrapes contract names from the definition and structurally builds the generic spec).

A later follow-up would enable generated the specs for ABI calls directly as well, and then also for importing behaviors from ACT.

### CODE REVIEW

File: kevm_pyk/src/kevm_pyk/solc_to_k.py [hunk 1]
```diff
@@ -16,23 +17,77 @@
...
+    if subprocess_res.returncode != 0:
+        fatal(f'solc error:\n{subprocess_res.stderr.decode()}')
 
     return json.loads(subprocess_res.stdout)
 
 
+def gen_spec_modules(kompiled_directory: Path):
+    kevm = KPrint(str(kompiled_directory))
+    kevm.symbolTable = kevmSymbolTable(kevm.symbolTable)
+    production_labels = [prod.klabel for module in kevm.definition.modules for prod in module.productions if prod.klabel is not None]
```
Comment (tothtamas28) [inline kevm_pyk/src/kevm_pyk/solc_to_k.py:53]: :sparkles: `KDefinition` defines `__iter__` in a way that you can just:

```suggestion
    production_labels = [prod.klabel for module in kevm.definition for prod in module.productions if prod.klabel is not None]
```

File: kevm_pyk/src/kevm_pyk/solc_to_k.py [hunk 2]
```diff
@@ -16,23 +17,77 @@
...
+    contract_names = [prod_label[9:] for prod_label in production_labels if prod_label.startswith('contract_')]
+    contract_function_labels = ['function_' + contract_name for contract_name in contract_names]
+    top_level_rules = [rule for module in kevm.definition.modules for rule in module.rules if type(rule.body) is KRewrite]
```
Comment (tothtamas28) [inline kevm_pyk/src/kevm_pyk/solc_to_k.py:56]:
```suggestion
    top_level_rules = [rule for module in kevm.definition for rule in module.rules if type(rule.body) is KRewrite]
```

File: kevm_pyk/src/kevm_pyk/solc_to_k.py [hunk 3]
```diff
@@ -16,23 +17,77 @@
...
+    contract_function_signatures = [rule.body.lhs for rule in top_level_rules if type(rule.body.lhs) is KApply and rule.body.lhs.label in contract_function_labels]
+    spec_modules = []
+    for contract_name in contract_names:
```
Comment (tothtamas28) [inline kevm_pyk/src/kevm_pyk/solc_to_k.py:59]: Consider extracting the body of this loop to a function.

File: kevm_pyk/src/kevm_pyk/solc_to_k.py [hunk 4]
```diff
@@ -54,16 +109,15 @@ def solc_to_k(kompiled_directory: Path, contract_file: Path, contract_name: str,
     function_sentences = generate_function_sentences(contract_name, contract_sort, abi)
     function_selector_alias_sentences = generate_function_selector_alias_sentences(contract_name, contract_sort, hashes)
 
-    binRuntimeProduction = KProduction(KSort('ByteArray'), [KTerminal('#binRuntime'), KTerminal('('), KNonTerminal(contract_sort), KTerminal(')')], att=KAtt({'klabel': 'binRuntime', 'alias': ''}))
-
-    contractProduction = KProduction(contract_sort, [KTerminal(contract_name)], att=KAtt({'klabel': f'contract_{contract_name}'}))
+    contractSubsort = KProduction(KSort('Contract'), [KNonTerminal(contract_sort)])
+    contractProduction = KProduction(contract_sort, [KTerminal(contract_name)], att=KAtt({'klabel': f'contract_{contract_name}', 'symbol': ''}))
```
Comment (tothtamas28) [inline kevm_pyk/src/kevm_pyk/solc_to_k.py:113]:
```suggestion
    contract_subsort = KProduction(KSort('Contract'), [KNonTerminal(contract_sort)])
    contract_production = KProduction(contract_sort, [KTerminal(contract_name)], att=KAtt({'klabel': f'contract_{contract_name}', 'symbol': ''}))
```

**Review by tothtamas28 — CHANGES_REQUESTED** [review]

File: kevm_pyk/src/kevm_pyk/solc_to_k.py (hunk 3 above)
Comment (ehildenb) [reply kevm_pyk/src/kevm_pyk/solc_to_k.py:87]: Done.

**Review by tothtamas28 — APPROVED** [review]

**Pull request merged**